src/webgpu/shader/execution/value_init.spec.ts

```diff
@@ -0,0 +1,370 @@
+  pass.setPipeline(pipeline);
+  pass.setBindGroup(0, bindGroup);
+  pass.dispatchWorkgroups(1);
+  pass.end();
+  t.queue.submit([encoder.finish()]);
+  t.expectGPUBufferValuesEqual(resultBuffer, new Uint32Array([0]));
+}
+
+g.test('scalars')
+  .desc(`Test that scalars in private, and function storage classes can be initialized to a value.`)
+  .params(u =>
+    u
+      .combine('addressSpace', ['private', 'function'] as const)
+      .combine('type', ['bool', 'f32', 'f16', 'i32', 'u32'] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const typeDecl = t.params.type;
+    const testValue = paramTypeToTestType(typeDecl);
+
+    const comparison = `if (testVar != ${testValue}) {
+      atomicStore(&output.failed, 1u);
+    }`;
+    const wgsl = generateShader(
+      t.params.type === 'f16',
+      t.params.addressSpace,
+      typeDecl,
+      testValue,
+      comparison
+    );
+
+    await run(t, wgsl);
+  });
+
+g.test('vec')
+  .desc(`Test that vectors in private, and function storage classes can be initialized to a value.`)
+  .params(u =>
+    u
+      .combine('addressSpace', ['private', 'function'] as const)
+      .combine('type', ['bool', 'f32', 'f16', 'i32', 'u32'] as const)
+      .combine('count', [2, 3, 4] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const typeDecl = `vec${t.params.count}<${t.params.type}>`;
+    const testValue = `${typeDecl}(${paramTypeToTestType(t.params.type)})`;
+
+    const comparison = `if (!all(testVar == ${testValue})) {
+      atomicStore(&output.failed, 1u);
+    }`;
+    const wgsl = generateShader(
+      t.params.type === 'f16',
+      t.params.addressSpace,
+      typeDecl,
+      testValue,
+      comparison
+    );
+
+    await run(t, wgsl);
+  });
+
+g.test('mat')
+  .desc(
+    `Test that matrices in private, and function storage classes can be initialized to a value.`
+  )
+  .params(u =>
+    u
+      .combine('addressSpace', ['private', 'function'] as const)
+      .combine('type', ['f32', 'f16'] as const)
+      .beginSubcases()
+      .combine('c', [2, 3, 4] as const)
+      .combine('r', [2, 3, 4] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const typeDecl = `mat${t.params.c}x${t.params.r}<${t.params.type}>`;
+    let testScalarValue = '';
+    switch (t.params.type) {
+      case 'f32':
+        testScalarValue = '5.0f';
+        break;
+      case 'f16':
+        testScalarValue = '5.0h';
+        break;
+    }
+
+    let testValue = `${typeDecl}(`;
+    [...Array(t.params.c)].map((_, i) => {
+      [...Array(t.params.r)].map((_, k) => {
+        testValue += `${testScalarValue},`;
+      });
+    });
```

Comment:
```suggestion
    for (let c = 0; c < t.params.c; c++) {
      for (let r = 0; r < t.params.r; r++) {
        testValue += `${testScalarValue},`;
      }
    }
```

Reply:
I've seen this go both ways, more C like and more functional-like.

Reply:
My objection to this is unnecessary complexity and performance.

* You're constructing an array that needs to be garbage collected.
* For each element of the array, you're calling a function and only using the index.
* That function does some logic, and doesn't return a value.
* The `undefined` value is then used to construct a second array, which is returned and not used. It's another thing we need to GC.

All of this _looks_ like it is doing something more complex than a for loop, when it isn't.